File: Resolução 8 & 9 PAA/colorir.py
N = 5  # Tamanho do tabuleiro (7x7)
cont = 0
C = 5

def verificacoes(board, x, y, color):
    for i in range(N):
        if board[x][i] == color or board[i][y] == color:
            return False

    # Verifica se a cor já está presente na diagonal principal
    for i in range(N):
        if board[i][i] == color:
            return False

    # Verifica se a cor já está presente na diagonal secundária
    for i in range(N):
        if board[i][N - 1 - i] == color:
            return False

    return True

def solveBoard(board, row, col):
    global cont
    if col == N:
        row += 1
        cont += 1
        col = 0

    if row == N:
        return True

    for color in range(1, C + 1):
        if verificacoes(board, row, col, color):
            board[row][col] = color

            if solveBoard(board, row, col + 1):
                return True
            board[row][col] = 0

    return False

if __name__ == "__main__":
    board = [[0] * N for _ in range(N)]  # Inicializa o tabuleiro com zeros
    if solveBoard(board, 0, 0):
        print("Tabuleiro colorido:")
        for i in range(N):
            print(*board[i])
        print(f"Número de soluções: {cont}")
    else:
        print("Não há solução.")

File: Resolução 8 & 9 PAA/colorir.cpp
#include <iostream>
#include <vector>

const int N = 6; // Tamanho do tabuleiro (6x6)

// Função para verificar se uma cor é válida para uma posição específica
bool isValidColor(const std::vector<std::vector<int>>& board, int row, int col, int color) {


    // Verifica se a cor é diferente das diagonais principais
    if (row == col && board[0][0] == color) {
        return false;
    }
    if (row + col == N - 1 && board[0][N - 1] == color) {
        return false;
    }

    // Verifica se a cor é adjacente a ela mesma
    if (row > 0 && board[row - 1][col] == color) {
        return false;
    }
    if (col > 0 && board[row][col - 1] == color) {
        return false;
    }

    return true;
}


// Função para encontrar a próxima posição vazia no tabuleiro
bool findNextEmptyPosition(const std::vector<std::vector<int>>& board, int& row, int& col) {
    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < N; ++j) {
            if (board[i][j] == 0) {
                row = i;
                col = j;
                return true;
            }
        }
    }
    return false;
}

// Função principal para resolver o problema de coloração do tabuleiro
int solveBoardColoring(std::vector<std::vector<int>>& board, int row, int col) {
    // Caso base: todas as posições preenchidas
    if (row == N) {
        // Verifique se a solução atende aos critérios (subtabuleiros, diagonais, etc.)
        // Se sim, retorne 1 (uma solução válida), caso contrário, retorne 0.
        // Exemplo simplificado (não verifica todas as condições):
        return 1;
    }

    // Encontre a próxima posição vazia
    int nextRow, nextCol;
    if (!findNextEmptyPosition(board, nextRow, nextCol)) {
        // Não há mais posições vazias, retorne 0.
        return 0;
    }

    int totalSolutions = 0;
    for (int color = 1; color <= 5; ++color) {
        if (isValidColor(board, nextRow, nextCol, color)) {
            board[nextRow][nextCol] = color;
            totalSolutions += solveBoardColoring(board, nextRow, nextCol);
            board[nextRow][nextCol] = 0; // Desfaz a atribuição para explorar outras cores
        }
    }

    return totalSolutions;
}

int main() {
    std::vector<std::vector<int>> board(N, std::vector<int>(N, 0)); // Inicializa o tabuleiro vazio
    int solutions = solveBoardColoring(board, 0, 0);
    std::cout << "Quantidade de soluções: " << solutions << std::endl;
    return 0;
}
